src/Algebra/Graph.hs

```diff
@@ -1187,68 +1288,69 @@ sparsifyKL n graph = KL.buildG (1, next - 1) ((n + 1, n + 2) : Exts.toList (res
-buildR g = g Empty Vertex Overlay Connect
-{-# INLINE [1] buildR #-}
+-- | Build a graph given an interpretation of the four graph construction primitives 'empty',
+-- 'vertex', 'overlay' and 'connect', in this order. See examples for further clarification.
+--
+-- Functions expressed with 'buildg' are good producers.
+--
+-- @
+-- buildg f                                                   == f 'empty' 'vertex' 'overlay' 'connect'
+-- buildg (\\e _ _ _ -> e)                                     == 'empty'
+-- buildg (\\_ v _ _ -> v x)                                   == 'vertex' x
+-- buildg (\\e v o c -> o ('foldg' e v o c x) ('foldg' e v o c y)) == 'overlay' x y
+-- buildg (\\e v o c -> c ('foldg' e v o c x) ('foldg' e v o c y)) == 'connect' x y
+-- buildg (\\e v o _ -> 'foldr' o e ('map' v xs))                  == 'vertices' xs
+-- buildg (\\e v o c -> 'foldg' e v o ('flip' c) g)                 == 'transpose' g
+-- 'foldg' e v o c (buildg f)                                   == f e v o c
+-- @
+buildg :: (forall b. b -> (a -> b) -> (b -> b -> b) -> (b -> b -> b) -> b) -> Graph a
+buildg f = f Empty Vertex Overlay Connect
+{-# INLINE [1] buildg #-}
 
 composeR :: (b -> c) -> (a -> b) -> a -> c
 composeR = (.)
-{-# INLINE [0] composeR #-}
+{-# INLINE [1] composeR #-}
 
 matchR :: b -> (a -> b) -> (a -> Bool) -> a -> b
 matchR e v p = \x -> if p x then v x else e
 {-# INLINE [0] matchR #-}
 
--- These rules transform functions into their buildR equivalents.
+-- These rules transform functions into their buildg equivalents.
 {-# RULES
-"buildR/bindR" forall f g.
-    bindR g f = buildR (\e v o c -> foldg e (composeR (foldg e v o c) f) o c g)
-
-"buildR/induce" [~1] forall p g.
-    induce p g = buildR (\e v o c -> foldg e (matchR e v p) o c g)
-
-"buildR/foldg(fc)" [~1] forall (f :: forall b. (b -> b -> b) -> (b -> b -> b)) g.
-    foldg Empty Vertex Overlay (f Connect) g = buildR (\e v o c -> foldg e v o (f c) g)
-
-"buildR/foldg(fo)" [~1] forall (f :: forall b. (b -> b -> b) -> (b -> b -> b)) g.
-    foldg Empty Vertex (f Overlay) Connect g = buildR (\e v o c -> foldg e v (f o) c g)
-
-"buildR/foldg(fo)(hc)" [~1] forall (f :: forall b. (b -> b -> b) -> (b -> b -> b)) (h :: forall b. (b -> b -> b) -> (b -> b -> b)) g.
-    foldg Empty Vertex (f Overlay) (h Connect) g = buildR (\e v o c -> foldg e v (f o) (h c) g)
+"buildg/induce" [~1] forall e v o c p g.
+    foldg e v o c (induce p g) = foldg e (matchR e v p) o c g
  #-}
 
 -- Rewrite rules for fusion.
 {-# RULES
--- Fuse a foldg followed by a buildR
-"foldg/buildR" forall e v o c (g :: Foldg a).
-    foldg e v o c (buildR g) = g e v o c
+-- Fuse a foldg followed by a buildg
+"foldg/buildg" forall e v o c (g :: forall b. b -> (a -> b) -> (b -> b -> b) -> (b -> b -> b) -> b).
+    foldg e v o c (buildg g) = g e v o c
 
 -- Fuse composeR's. This occurs when two adjacent 'bindR' were rewritted into
--- their buildR form.
+-- their buildg form.
 "bindR/bindR" forall c f g.
```

Comment:
Why is this rule called `bindR/bindR`? I guess it should be `composeR/composeR`.